---
title: "mNSC Bulk RNA-Seq Analysis"
subtitle: "A Modular and Reproducible Pipeline"
author: "Ryan Miller"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
output:
  html_document:
    toc: yes
    toc_float: no
    toc_depth: 5
    code_folding: show
    theme: united
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = TRUE, warning = TRUE)
```

# Project 250902_mNSC
This document is the central hub for a modular DESeq2 RNA-seq analysis pipeline.

---

# 1. Setup and Configuration

## 1.1 Load Child Scripts
```{r child-script-setup, file='scripts/00_setup.R', echo=TRUE}
```
```{r child-script-genes, file='scripts/01_genes_mouse.R', echo=TRUE}
```
```{r child-script-qc, file='scripts/02_QC.R', echo=TRUE}
```
```{r child-script-deseq, file='scripts/03_DESeq2.R', echo=TRUE}
```

## 1.2 Project Setup and Data Loading
```{r project-setup-and-caching}
RUN_ANALYSIS <- TRUE
cfg <- list(
  date = format(Sys.Date(), "%y%m%d"),
  project_name = "250902_mNSC",
  dir_input = here::here("input"),
  dir_output = here::here("output"),
  dir_sf = here::here("sf"),
  dir_genesets = here::here("genesets"),
  dir_graphs_png = here::here("graphs", "png"),
  dir_graphs_pdf = here::here("graphs", "pdf"),
  dir_graphs_png_kinases = here::here("graphs", "png", "kinases"),
  dir_graphs_pdf_kinases = here::here("graphs", "pdf", "kinases"),
  dir_results_csv = here::here("output", "results_csv"),
  file_samples = here::here("input", "samples.csv"),
  file_gtf = here::here("input", "gencode.vM37.primary_assembly.annotation.gtf.gz"),
  file_kinases = here::here("genesets", "201006_composite_kinases.csv"),
  qval_threshold = 0.05,
  lfc_threshold = 1.0,
  filter_min_count = 10,
  batch_vars = c("SeqBatch"),
  main_vars = c("Driver", "Model", "Host"),
  deg_design_vars = c("Host", "Model", "Driver"),
  interaction_vars = c("Model", "Driver"),
  filter_by_factor = "Driver",
  qc_plot_settings = list(project_genes = c("Egfr", "Pdgfra", "Pten", "Cdkn2a")),
  ryb = colorRampPalette(rev(
    RColorBrewer::brewer.pal(n = 9, name = "RdYlBu")
  ))(100),
  analysis_definitions = list(
    sample_subsets = list(
      Full_Dataset = list(),
      NS1_Model = list(Model = "NS1"),
      BL6_Host = list(Host = "BL6")
    ),
    gene_filters = list(
      all_genes = list(filter_gene_types = NULL),
      protein_coding = list(filter_gene_types = "protein_coding")
    ),
    model_types = c("additive", "interaction")
  )
)

RUN_STYLER <- TRUE
if (RUN_STYLER) {
  files_to_style <- c(here::here(paste0(cfg$project_name, ".Rmd")),
                      list.files(
                        here::here("scripts"),
                        pattern = "\\.R$",
                        full.names = TRUE
                      ))
  styler::style_file(files_to_style)
}

if (RUN_ANALYSIS == FALSE) {
  results_file_path <- here::here(cfg$dir_input, paste0(cfg$project_name, "_results.RData"))
  if (file.exists(results_file_path)) {
    load(results_file_path, envir = .GlobalEnv)
    RUN_ANALYSIS <- FALSE
  }
}

samples <- read.csv(cfg$file_samples, row.names = 1, header = TRUE)
validate_config_and_inputs(cfg, samples)
kinase_genes <- if (file.exists(cfg$file_kinases)) toupper(unique(read.csv(cfg$file_kinases)$Mouse_symbol)) else NULL
```

---

# 2. Computational Analysis
```{r computational-analysis, echo=FALSE, eval=RUN_ANALYSIS}
set.seed(42)
lapply(
  c(
    cfg$dir_output,
    cfg$dir_graphs_png,
    cfg$dir_graphs_pdf,
    cfg$dir_results_csv,
    cfg$dir_graphs_png_kinases,
    cfg$dir_graphs_pdf_kinases
  ),
  dir.create,
  recursive = TRUE,
  showWarnings = FALSE
)

samples[, sapply(samples, is.character)] <- lapply(samples[, sapply(samples, is.character)], as.factor)
files_sf <- file.path(cfg$dir_sf, paste0(rownames(samples), ".sf"))
names(files_sf) <- rownames(samples)

all_dds_objects <- list()
all_results_lists <- list()

for (s_name in names(cfg$analysis_definitions$sample_subsets)) {
  for (filt_name in names(cfg$analysis_definitions$gene_filters)) {
    filt_params <- cfg$analysis_definitions$gene_filters[[filt_name]]
    gtf_maps <- create_tx2gene_maps(cfg$file_gtf,
                                    cfg$dir_input,
                                    filt_params$filter_gene_types)
    
    for (m_type in cfg$analysis_definitions$model_types) {
      analysis_name <- paste(s_name, filt_name, m_type, sep = "_")
      message(paste("\n--- STARTING ANALYSIS FOR:", analysis_name, "---"))
      
      current_samples <- samples
      subset_criteria <- cfg$analysis_definitions$sample_subsets[[s_name]]
      if (length(subset_criteria) > 0) {
        for (col in names(subset_criteria)) {
          values_to_keep <- subset_criteria[[col]]
          current_samples <- current_samples[current_samples[[col]] %in% values_to_keep, ]
        }
        current_samples <- droplevels(current_samples)
      }
      
      if (nrow(current_samples) == 0) {
        message(
          paste(
            "...SKIPPING analysis for",
            analysis_name,
            "because the data subset was empty."
          )
        )
        next
      }
      
      current_files_sf <- files_sf[rownames(current_samples)]
      relevant_vars <- names(which(sapply(current_samples, function(col)
        length(unique(col)) > 1)))
      batch_terms <- intersect(cfg$batch_vars, relevant_vars)
      
      if (m_type == "additive") {
        design_terms <- intersect(cfg$deg_design_vars, relevant_vars)
        design_str <- paste("~", paste(c(batch_terms, design_terms), collapse = " + "))
      } else {
        main_terms <- intersect(cfg$main_vars, relevant_vars)
        if (all(cfg$interaction_vars %in% main_terms)) {
          other_main_terms <- setdiff(main_terms, cfg$interaction_vars)
          design_str <- paste("~", paste(c(
            batch_terms,
            other_main_terms,
            paste(cfg$interaction_vars, collapse = " * ")
          ), collapse = " + "))
        } else {
          design_terms <- intersect(cfg$deg_design_vars, relevant_vars)
          design_str <- paste("~", paste(c(batch_terms, design_terms), collapse = " + "))
        }
      }
      
      dds_processed <- run_deseq_core(
        current_samples,
        current_files_sf,
        as.formula(design_str),
        gtf_maps$tx2gene_pcg,
        cfg$filter_by_factor,
        cfg$filter_min_count
      )
      all_dds_objects[[analysis_name]] <- dds_processed
      generate_qc_plots(dds_processed,
                        analysis_name,
                        gtf_maps$gene_name_map,
                        cfg)
      
      factors_to_compare <- intersect(cfg$main_vars, relevant_vars)
      comparisons <- unlist(lapply(factors_to_compare, function(f) {
        if (length(levels(current_samples[[f]])) > 1) {
          combn(levels(current_samples[[f]]), 2, function(p)
            list(list(
              factor = f,
              group1 = p[1],
              group2 = p[2]
            )), simplify = FALSE)
        }
      }), recursive = FALSE)
      all_results_lists[[analysis_name]] <- extract_comparisons(
        dds_processed,
        comparisons,
        analysis_name,
        gtf_maps$gene_name_map,
        cfg,
        kinase_genes
      )
    }
  }
}
save(
  cfg,
  samples,
  gtf_maps,
  all_dds_objects,
  all_results_lists,
  kinase_genes,
  file = here::here(cfg$dir_input, paste0(cfg$project_name, "_results.RData"))
)
```

---

# 3. Report Generation
## 3.1 Sample Metadata
```{r sample-metadata-table}
reactable::reactable(
  samples,
  searchable = TRUE,
  filterable = TRUE,
  bordered = TRUE,
  highlight = TRUE
)
```
## 3.2 Quality Control
```{r dynamic-qc-plots, results='asis'}
cat("\n### PCA Plots\n")
for (analysis_name in names(all_dds_objects)) {
  cat(paste("\n####", gsub("_", " ", analysis_name), "\n"))
  cat(paste0("![](", here::here(
    cfg$dir_graphs_png,
    paste0(cfg$date, "_", analysis_name, "_PCA.png")
  ), ")"))
}

cat("\n### Sample Distance Heatmaps\n")
for (analysis_name in names(all_dds_objects)) {
  cat(paste("\n####", gsub("_", " ", analysis_name), "\n"))
  cat(paste0("![](", here::here(
    cfg$dir_graphs_png,
    paste0(cfg$date, "_", analysis_name, "_Heatmap.png")
  ), ")"))
}
```
## 3.3 Gene of Interest Expression
```{r goi-plots, results='asis'}
for (analysis_name in names(all_dds_objects)) {
  cat(paste("\n### GOI Expression for", gsub("_", " ", analysis_name), "\n"))
  for (gene_name in cfg$qc_plot_settings$project_genes) {
    plot_path <- here::here(
      cfg$dir_graphs_png,
      paste0(
        cfg$date,
        "_",
        analysis_name,
        "_GenePlot_",
        gene_name,
        ".png"
      )
    )
    if (file.exists(plot_path))
      cat(paste0("\n![](", plot_path, ")"))
  }
}
```
## 3.4 Differential Expression Summary
```{r results-summary-table}
summary_list <- lapply(names(all_results_lists), function(analysis_name) {
  analysis_results <- all_results_lists[[analysis_name]]
  if (length(analysis_results) == 0)
    return(NULL)
  
  lapply(names(analysis_results), function(comp_name) {
    res_df <- analysis_results[[comp_name]]
    sig_df <- res_df %>% dplyr::filter(padj < cfg$qval_threshold &
                                         abs(log2FoldChange) > cfg$lfc_threshold)
    data.frame(
      Comparison = comp_name,
      `Total Genes` = nrow(res_df),
      DEGs = nrow(sig_df),
      Up = sum(sig_df$log2FoldChange > 0),
      Down = sum(sig_df$log2FoldChange < 0),
      check.names = FALSE
    )
  })
})

summary_df <- dplyr::bind_rows(unlist(summary_list, recursive = FALSE))

if (!is.null(summary_df) && nrow(summary_df) > 0) {
  reactable::reactable(
    summary_df,
    searchable = TRUE,
    bordered = TRUE,
    highlight = TRUE,
    columns = list(Comparison = colDef(minWidth = 400))
  )
} else {
  cat("\n*No valid comparisons were found to summarize.*\n")
}
```
## 3.5 Detailed Comparison Results
```{r dynamic-results-report, results='asis'}
for (analysis_name in names(all_results_lists)) {
  cat(paste("\n### DEG Results for", gsub("_", " ", analysis_name), "\n"))
  analysis_results <- all_results_lists[[analysis_name]]
  if (length(analysis_results) == 0)
    next
  for (comp_name in names(analysis_results)) {
    cat(paste("\n####", gsub("_", " ", comp_name), "\n"))
    res_df <- analysis_results[[comp_name]]
    sig_genes <- res_df %>% dplyr::filter(padj < cfg$qval_threshold &
                                            abs(log2FoldChange) > cfg$lfc_threshold & !is.na(padj))
    if (nrow(sig_genes) > 0) {
      print(
        reactable::reactable(
          sig_genes %>%
            dplyr::select(
              gene_name,
              human_ortholog,
              baseMean,
              log2FoldChange,
              padj
            ) %>%
            dplyr::arrange(padj),
          defaultPageSize = 5,
          searchable = TRUE,
          bordered = TRUE
        )
      )
    }
    
    cat("\n\n##### Volcano Plot\n")
    cat(paste0("![](", here::here(
      cfg$dir_graphs_png,
      paste0(cfg$date, "_", comp_name, "_volcano.png")
    ), ")"))
    cat("\n\n##### Top 50 DEG Heatmap\n")
    cat(paste0("![](", here::here(
      cfg$dir_graphs_png,
      paste0(cfg$date, "_", comp_name, "_DEG_heatmap_top50.png")
    ), ")"))
    cat("\n\n##### Top 250 DEG Heatmap\n")
    path_250 <- here::here(
      cfg$dir_graphs_png,
      paste0(cfg$date, "_", comp_name, "_DEG_heatmap_top250.png")
    )
    if (file.exists(path_250))
      cat(paste0("![](", path_250, ")"))
    
    cat("\n\n##### Top Kinase DEG Heatmap\n")
    path_kinase_heatmap <- here::here(
      cfg$dir_graphs_png,
      paste0(cfg$date, "_", comp_name, "_Kinase_DEG_heatmap.png")
    )
    if (file.exists(path_kinase_heatmap))
      cat(paste0("![](", path_kinase_heatmap, ")"))
    
    cat("\n\n##### DE Kinase Box Plots\n")
    sig_kinases <- sig_genes %>% dplyr::filter(gene_name_upper %in% kinase_genes |
                                                 toupper(human_ortholog) %in% kinase_genes)
    if (nrow(sig_kinases) > 0) {
      for (kinase_name in sig_kinases$gene_name) {
        plot_path <- here::here(
          cfg$dir_graphs_png_kinases,
          paste0(
            cfg$date,
            "_",
            comp_name,
            "_Boxplot_",
            kinase_name,
            ".png"
          )
        )
        if (file.exists(plot_path))
          cat(paste0("![](", plot_path, ")"))
      }
    } else {
      cat("\n*No DE kinases found.*\n")
    }
  }
}
```

---

# 4. Session
```{r session-info}
sessionInfo()
```
